File: crates/xyd_openapi2opencli/src/tree.rs
//! CommandTree — port of tree.ts. Static path segments become nested resource
//! nodes; leaves attach to their node, deduped with `-N`, and emit sorted
//! (children by localeCompare, leaves by action rank then localeCompare).

use crate::jsrt::locale_compare;
use crate::model::Command;

fn leaf_rank(name: &str) -> i32 {
    match name {
        "list" => 0,
        "create" => 1,
        "retrieve" | "get" => 2,
        "update" | "modify" | "replace" => 3,
        "delete" => 4,
        _ => 100,
    }
}

struct Node {
    name: String,
    children: Vec<Node>,
    leaves: Vec<Command>,
    used_leaf_names: std::collections::HashSet<String>,
}

impl Node {
    fn new(name: &str) -> Self {
        Node {
            name: name.to_string(),
            children: Vec::new(),
            leaves: Vec::new(),
            used_leaf_names: Default::default(),
        }
    }
}

pub struct CommandTree {
    root: Node,
}

impl Default for CommandTree {
    fn default() -> Self {
        Self::new()
    }
}

impl CommandTree {
    pub fn new() -> Self {
        CommandTree {
            root: Node::new(""),
        }
    }

    pub fn insert(&mut self, resource_path: &[String], mut leaf: Command) {
        let mut node = &mut self.root;
        for seg in resource_path {
            let pos = node.children.iter().position(|c| &c.name == seg);
            let idx = match pos {
                Some(i) => i,
                None => {
                    node.children.push(Node::new(seg));
                    node.children.len() - 1
                }
            };
            node = &mut node.children[idx];
        }
        let mut name = leaf.name.clone();
        if node.used_leaf_names.contains(&name) {
            let mut i = 2u32;
            while node
                .used_leaf_names
                .contains(&format!("{}-{}", leaf.name, i))
            {
                i += 1;
            }
            name = format!("{}-{}", leaf.name, i);
            leaf.name = name.clone();
        }
        node.used_leaf_names.insert(name);
        node.leaves.push(leaf);
    }

    pub fn emit(&self) -> Vec<Command> {
        emit_children(&self.root)
    }
}

fn emit_children(node: &Node) -> Vec<Command> {
    let mut children: Vec<&Node> = node.children.iter().collect();
    children.sort_by(|a, b| locale_compare(&a.name, &b.name));
    let mut out: Vec<Command> = children.iter().map(|c| emit_node(c)).collect();

    let mut leaves = node.leaves.clone();
    leaves.sort_by(|a, b| {
        leaf_rank(&a.name)
            .cmp(&leaf_rank(&b.name))
            .then_with(|| locale_compare(&a.name, &b.name))
    });
    out.extend(leaves);
    out
}

fn emit_node(node: &Node) -> Command {
    let subs = emit_children(node);
    Command {
        name: node.name.clone(),
        commands: if subs.is_empty() { None } else { Some(subs) },
        ..Default::default()
    }
}

File: packages/xyd-openapi2opencli/src/impl-js/tree.ts
import type { Command } from '@xyd-js/opencli';

const ACTION_RANK: Record<string, number> = {
  list: 0,
  create: 1,
  retrieve: 2,
  get: 2,
  update: 3,
  modify: 3,
  replace: 3,
  delete: 4,
};

function leafRank(name: string): number {
  return name in ACTION_RANK ? ACTION_RANK[name] : 100;
}

interface TreeNode {
  name: string;
  children: Map<string, TreeNode>;
  leaves: Command[];
  usedLeafNames: Set<string>;
}

function newNode(name: string): TreeNode {
  return { name, children: new Map(), leaves: [], usedLeafNames: new Set() };
}

/**
 * Builds the OpenCLI command tree from operations. Static path segments become
 * nested resource nodes; leaf actions are attached to their resource node.
 */
export class CommandTree {
  private root = newNode('');

  insert(resourcePath: string[], leaf: Command): void {
    let node = this.root;
    for (const seg of resourcePath) {
      let child = node.children.get(seg);
      if (!child) {
        child = newNode(seg);
        node.children.set(seg, child);
      }
      node = child;
    }
    // Deduplicate leaf action names within the resource node.
    let name = leaf.name;
    if (node.usedLeafNames.has(name)) {
      let i = 2;
      while (node.usedLeafNames.has(`${name}-${i}`)) i++;
      name = `${name}-${i}`;
      leaf.name = name;
    }
    node.usedLeafNames.add(name);
    node.leaves.push(leaf);
  }

  emit(): Command[] {
    return this.emitChildren(this.root);
  }

  private emitChildren(node: TreeNode): Command[] {
    const children = [...node.children.values()]
      .sort((a, b) => a.name.localeCompare(b.name))
      .map((child) => this.emitNode(child));
    const leaves = [...node.leaves].sort(
      (a, b) => leafRank(a.name) - leafRank(b.name) || a.name.localeCompare(b.name),
    );
    return [...children, ...leaves];
  }

  private emitNode(node: TreeNode): Command {
    const cmd: Command = { name: node.name };
    const subs = this.emitChildren(node);
    if (subs.length) cmd.commands = subs;
    return cmd;
  }
}
